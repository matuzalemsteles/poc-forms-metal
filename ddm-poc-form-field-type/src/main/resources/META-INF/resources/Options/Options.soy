{namespace Options}

{template .render}
    {@param? _handleTextChange: any}
    {@param? editable: bool}
    {@param? helpText: string}
    {@param? items: list<[
        checked: bool,
        disabled: bool,
        id: string,
        inline: bool,
        label: string,
        name: string,
        showLabel: bool,
        value: string
    ]>}
    {@param? id: string}
    {@param? label: string}
    {@param? required: bool}
    {@param? showLabel: bool}
    {@param? spritemap: string}
    {@param? placeholder: string}

    {call FieldBase.render}
        {param contentRenderer kind="html"}
            {if $items}
                <div class="field-options">
                    {for $option in $items}
                        {let $fieldConfig kind="html"}
                            {call Text.render}
                                {param events: [
                                    'fieldEdit': $_handleTextChange
                                ] /}
                                {param editable: $editable ? $editable : $option.disabled /}
                                {param value: $option.label and $option.value ? $option.label : $placeholder /}
                                {param showLabel: false /}
                                {param ref: 'text' /}
                            {/call}
                        {/let}

                        {$fieldConfig}

                        {if isLast($option)}
                            {call Text.render}
                                {param events: [
                                    'fieldEdit': $_handleTextChange
                                ] /}

                                {param editable: $editable ? $editable : $option.disabled /}
                                {param placeholder: $placeholder /}
                                {param showLabel: false /}
                                {param ref: 'textVoid' /}
                            {/call}
                        {/if}
                    {/for}
                </div>
            {/if}
        {/param}

        {param helpText: $helpText /}
        {param id: $id /}
        {param label: $label /}
        {param required: $required /}
        {param showLabel: $showLabel /}
        {param spritemap: $spritemap /}
    {/call}
{/template}