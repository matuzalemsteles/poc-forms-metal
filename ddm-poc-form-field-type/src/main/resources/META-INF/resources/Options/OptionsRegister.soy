{namespace OptionsRegister}

{deltemplate LayoutRenderer.RegisterFieldType variant="'options'"}
    {call Options.render data="all" /}
{/deltemplate}