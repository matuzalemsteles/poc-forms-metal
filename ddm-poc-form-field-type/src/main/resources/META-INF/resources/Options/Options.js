import {Config} from 'metal-state';
import Component from 'metal-component';
import dom from 'metal-dom';
import FieldBase from 'ddm-poc-form-field-type/FieldBase/index.js';
import Text from 'ddm-poc-form-field-type/Text/index.js';
import Soy from 'metal-soy';

import templates from './Options.soy.js';
import OptionsRegister from './OptionsRegister.soy.js';

/**
 * Options.
 * @extends Component
 */
class Options extends Component {
    static STATE = {
        /**
         * @default false
         * @instance
         * @memberof Options
         * @type {?bool}
         */
        editable: Config.bool().value(false),

        /**
         * @default undefined
         * @instance
         * @memberof Options
         * @type {?(string|undefined)}
         */
        helpText: Config.string(),

        /**
         * @default undefined
         * @instance
         * @memberof Options
         * @type {?(string|undefined)}
         */
        items: Config.arrayOf(Config.shapeOf({
            checked: Config.bool().value(false),
            disabled: Config.bool().value(false),
            id: Config.string(),
            inline: Config.bool().value(false),
            label: Config.string(),
            name: Config.string(),
            showLabel: Config.bool().value(true),
            value: Config.string()
        })).internal(),

        /**
         * @default undefined
         * @instance
         * @memberof Options
         * @type {?(string|undefined)}
         */
        id: Config.string(),

        /**
         * @default undefined
         * @instance
         * @memberof Options
         * @type {?(string|undefined)}
         */
        label: Config.string(),

        /**
         * @default false
         * @instance
         * @memberof Options
         * @type {?bool}
         */
        required: Config.bool().value(false),

        /**
         * @default true
         * @instance
         * @memberof Options
         * @type {?bool}
         */
        showLabel: Config.bool().value(true),

        /**
         * @default undefined
         * @instance
         * @memberof Options
         * @type {?(string|undefined)}
         */
        spritemap: Config.string(),

        key: Config.string(),

        placeholder: Config.string()
    }

    _getFieldIndex(element) {
        return Array.prototype.indexOf.call(
            Array.prototype.filter.call(
				element.parentElement.children,
				childrenElement =>
					childrenElement.className === 'form-group'
			),
            element
        );
    }

    _handleTextChange(data) {
        const { value, originalEvent } = data;
        const { key } = this;
        const fieldIndex = this._getFieldIndex(originalEvent.delegateTarget.parentNode);

        if (typeof this.items[fieldIndex] === 'undefined') {
            const newItem = {'label': value};
            this.items.push(newItem);
        } else {
            this.items[fieldIndex].label = value;
        }

        this.setState({items: this.items});

        this.emit('fieldEdit', {
            value: this.items,
            key,
            originalEvent: originalEvent
        });

    }
}

Soy.register(Options, templates);

export default Options;