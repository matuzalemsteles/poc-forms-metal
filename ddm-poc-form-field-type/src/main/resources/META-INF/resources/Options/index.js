import Options from './Options.es';

export default Options;