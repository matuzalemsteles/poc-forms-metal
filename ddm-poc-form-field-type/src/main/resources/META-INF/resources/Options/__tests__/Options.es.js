import Options from '../Options.es';
import {dom as MetalTestUtil} from 'metal-dom';

let component;
let spritemap = 'icons.svg';

describe('Options', () => {
    afterEach(() => {
        if (component) {
			component.dispose();
        }
    });

    it('should be not edidable', () => {
        component = new Options({
            editable: false,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a helptext', () => {
        component = new Options({
            helpText: 'Type something',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should render items', () => {
        component = new Options({
            items: [{
                checked: false,
                disabled: false,
                id: 'id',
                inline: false,
                label: 'label',
                name: 'name',
                showLabel: true,
                value: 'item'
            }]
        });

        expect(component).toMatchSnapshot();
    });

    it('should render no items when items is empty', () => {
        component = new Options({
            items: []
        });

        expect(component).toMatchSnapshot();
    });

    it('should have an id', () => {
        component = new Options({
            id: 'ID',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a label', () => {
        component = new Options({
            label: 'label',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should not be required', () => {
        component = new Options({
            required: false,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should render Label if showLabel is true', () => {
        component = new Options({
            label: 'text',
            showLabel: true,
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a spritemap', () => {
        component = new Options({
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a key', () => {
        component = new Options({
            key: 'key',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should have a placeholder', () => {
        component = new Options({
            placeholder: 'Placeholder',
            spritemap: spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should emit a field edit event on any item value change', () => {
        const handleFieldEdit = jest.fn();
        const events = {fieldEdit: handleFieldEdit};
        component = new Options({
            items: [
                {
                    checked: false,
                    disabled: false,
                    id: 'id',
                    inline: false,
                    label: 'label',
                    name: 'name',
                    showLabel: true,
                    value: 'item'
                }
            ],
            spritemap: spritemap,
            events
        });

        MetalTestUtil.triggerEvent(component.element.querySelector('input'), 'input', {});

        expect(handleFieldEdit).toHaveBeenCalled();
    });

    it('should add a new item when add a content in the last option', () => {
        const handleFieldEdit = jest.fn();
        const events = {fieldEdit: handleFieldEdit};
        component = new Options({
            items: [
                {
                    checked: false,
                    disabled: false,
                    id: 'id',
                    inline: false,
                    label: 'label',
                    name: 'name',
                    showLabel: true,
                    value: 'item'
                }
            ],
            spritemap: spritemap,
            events
        });
        const initialSize = component.items.length;
        MetalTestUtil.triggerEvent(component.element.querySelector('.field-options .form-group:last-child input'), 'input', {});
        const finalSize = component.items.length;

        expect(finalSize > initialSize).toBe(true);
    });

    it('should propagate the field edit event when any item value change', () => {
        component = new Options({
            items: [
                {
                    checked: false,
                    disabled: false,
                    id: 'id',
                    inline: false,
                    label: 'label',
                    name: 'name',
                    showLabel: true,
                    value: 'item'
                }
            ],
            spritemap: spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        MetalTestUtil.triggerEvent(component.element.querySelector('input'), 'input', {});

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'fieldEdit',
			expect.any(Object)
		);
    });
});