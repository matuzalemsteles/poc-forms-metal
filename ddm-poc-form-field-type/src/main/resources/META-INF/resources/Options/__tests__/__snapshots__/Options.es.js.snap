// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options should be not edidable 1`] = `
<div class="form-group">
  <label></label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should have a helptext 1`] = `
<div class="form-group">
  <label></label>
  <span class="form-text">Type something</span>
</div>
`;

exports[`Options should have a key 1`] = `
<div class="form-group">
  <label></label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should have a label 1`] = `
<div class="form-group">
  <label>label </label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should have a placeholder 1`] = `
<div class="form-group">
  <label></label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should have a spritemap 1`] = `
<div class="form-group">
  <label></label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should have an id 1`] = `
<div class="form-group">
  <label for="ID"></label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should not be required 1`] = `
<div class="form-group">
  <label></label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should render Label if showLabel is true 1`] = `
<div class="form-group">
  <label>text </label>
  <span class="form-text"></span>
</div>
`;

exports[`Options should render items 1`] = `
<div class="form-group">
  <label></label>
  <div class="field-options">
    <div class="form-group">
      <label></label>
      <input class="form-control" type="text" disabled="disabled" value="label">
      <span class="form-text"></span>
    </div>
    <div class="form-group">
      <label></label>
      <input class="form-control" type="text" disabled="disabled">
      <span class="form-text"></span>
    </div>
  </div>
  <span class="form-text"></span>
</div>
`;

exports[`Options should render no items when items is empty 1`] = `
<div class="form-group">
  <label></label>
  <div class="field-options"></div>
  <span class="form-text"></span>
</div>
`;
